src/sqlfluff/rules/structure/ST06.py

```diff
@@ -86,35 +86,34 @@ def _eval(self, context: RuleContext) -> EvalResultType:
         ]  # type: ignore
 
         assert context.segment.is_type("select_clause")
-        # Ignore select clauses which belong to:
-        # - set expression, which is most commonly a union
-        # - insert_statement
-        # - create table statement
-        #
-        # In each of these contexts, the order of columns in a select should
-        # be preserved.
-        if len(context.parent_stack) >= 2 and context.parent_stack[-2].is_type(
-            "insert_statement", "set_expression"
-        ):
-            return None
-        if (
-            len(context.parent_stack) >= 3
-            and context.parent_stack[-3].is_type("insert_statement", "set_expression")
-            and context.parent_stack[-2].is_type("with_compound_statement")
-        ):
-            return None
-        if len(context.parent_stack) >= 3 and context.parent_stack[-3].is_type(
-            "create_table_statement", "merge_statement"
-        ):
-            return None
-        if (
-            len(context.parent_stack) >= 4
-            and context.parent_stack[-4].is_type(
-                "create_table_statement", "merge_statement"
-            )
-            and context.parent_stack[-2].is_type("with_compound_statement")
-        ):
-            return None
+
+        # insert, merge, create table, union are order-sensitive
+        for seg in reversed(context.parent_stack):
+            if seg.is_type(
+                "insert_statement",
+                "set_expression",
+                "create_table_statement",
+                "merge_statement",
+            ):
+                return None
+
+        # CTE is order-sensitive only if CTE is referenced in set expression
+        for seg in reversed(context.parent_stack):
+            if seg.is_type("common_table_expression"):
+                cte_identifier = seg.get_child("identifier")
+                assert cte_identifier is not None
+                maybe_with_compound_statement = seg.get_parent()
+                if maybe_with_compound_statement is None:
+                    break
+                with_compound_statement, _ = maybe_with_compound_statement
```

Comment:
The coverage checks are failing because of this line, but it's _very_ unlikely that would ever get used. If you add a comment here saying that clause is mostly a safety measure, then I think it's also ok to mark it as `# pragma: no cover`.

It would only be `None` if a fix has modified it in some way and we haven't yet re-populated the parent. Even in that scenario it's very unlikely.